File: src/com/imob/fuzzy/Fuzzy.java
/**
 * File: Fuzzy.java
 * This file implements a fuzzy logic for a specialist real state seller
 * Created by 
 *      Jonathan Arantes
 *      Rubia Marques
 *      Ana Paula
 * v0.1
 */

package com.imob.fuzzy;

import java.util.ArrayList;

public class Fuzzy {
    public static final long _INFINITY =  9000000;

    public float min (float l1, float l2) {
        if(l1 < l2) {
            return l1;
        } else {
            return l2;
        }
    }

    public float min (ArrayList<Float> l) {
        float m = _INFINITY;
        for (int i = 0; i < l.size(); i++) {
            if (l.get(i) < m) {
                m = l.get(i);
            }
        }
        return m;
    }

    public float max (float l1, float l2) {
        if(l1 > l2) {
            return l1;
        } else {
            return l2;
        }
    }

    public float max (ArrayList<Float> l) {
        float m = _INFINITY;
        for (int i = 0; i < l.size(); i++) {
            if (l.get(i) > m) {
                m = l.get(i);
            }
        }
        return m;
    }

    public ArrayList<Float> equals (ArrayList<Float> l1, ArrayList<Float> l2) {
        ArrayList<Float> resp = new ArrayList<>();
        for (int i = 0; i < l1.size(); i++) {
            if(l1.get(i) == l2.get(i)) {
                resp.add(l1.get(i));
            }
        }
        return resp;
    }

    public ArrayList<Float> complement (ArrayList<Float> l) {
        ArrayList<Float> resp = new ArrayList<>();
        for (int i = 0; i < l.size(); i++) {
            resp.add(1 - l.get(i));
        }
        return resp;
    }

    public boolean empty (ArrayList<Float> l) {
        if(l == null || l.size() == 0) {
            return true;
        }
        return false;
    }

    public ArrayList<Float> union (ArrayList<Float> l1, ArrayList<Float> l2) {
        ArrayList<Float> resp = new ArrayList<>();
        for(int i = 0; i < l1.size(); i++) {
            resp.add( max(l1.get(i), l2.get(i)) );
        }
        return resp;
    }

    public ArrayList<Float> intersection (ArrayList<Float> l1, ArrayList<Float> l2) {
        ArrayList<Float> resp = new ArrayList<>();
        for(int i = 0; i < l1.size(); i++) {
            resp.add( min(l1.get(i), l2.get(i)) );
        }
        return resp;
    }

    public boolean subconjunct (ArrayList<Float> l1, ArrayList<Float> l2) {
        return l1.containsAll(l2);
    }

    public ArrayList<Float> trunc(ArrayList<Float> l) {
        ArrayList<Float> resp = new ArrayList<>();
        for (int i = 0; i < l.size(); i++) {
            if( l.get(i) >= 1 ) {
                resp.add((float)1);
            } else if ( l.get(i) <= 0 ) {
                resp.add((float)0);
            } else {
                resp.add(l.get(i));
            }
        }
        return resp;
    }

    public Float trunc(Float l) {
        if( l > 1 ) {
            return (float)1;
        } else if ( l < 1 ) {
            return (float)0;
        } else {
            return l;
        }
    }
}

File: TypeScript/Fuzzy.ts
class Fuzzy {
    static _INFINITY: number = Number.MAX_VALUE;

    constructor() {}

    static min(l1: number, l2: number): number {
        if (l1 < l2) {
            return l1;
        } else {
            return l2;
        }
    }

    static minList(l: Array<number>): number {
        let m: number = Fuzzy._INFINITY;
        for (var i in l) {
            if (m > +i) {
                m = +i;
            }
        }
        return m;
    }

    static max(l1: number, l2: number): number {
        if (l1 > l2) {
            return l1;
        } else {
            return l2;
        }
    }

    static maxList(l: Array<number>): number {
        let m: number = 0;
        for (var i in l) {
            if (m < +i) {
                m = +i;
            }
        }
        return m;
    }

    static equals(l1: Array<number>, l2: Array<number>): Array<number> {
        let resp: Array<number>;
        for (let i = 0; i < l1.length; i++) {
            if(l1[i] == l2[i]) {
                resp.push(l1[i]);
            }
        }
        return resp;
    }

    static complement(l: Array<number>): Array<number> {
        let resp: Array<number>;
        for (let i = 0; i < l.length; i++) {
            resp.push(1 - l[i]);
        }
        return resp;
    }

    static isEmpty(l: Array<number>): boolean {
        if(l == null || l.length == 0) {
            return true;
        }
        return false;
    }
    
    static union(l1: Array<number>, l2: Array<number>): Array<number> {
        let resp: Array<number>;
        for (let i = 0; i < l1.length; i++) {
            resp.push( this.max(l1[i], l2[i]) );
        }
        return resp;
    }

    static intersection(l1: Array<number>, l2: Array<number>): Array<number> {
        let resp: Array<number>;
        for (let i = 0; i < l1.length; i++) {
            resp.push( this.min(l1[i], l2[i]) );
        }        
        return resp;
    }

    static subconjunct(l1: Array<number>, l2: Array<number>): Array<number> {
        let _l1 = l1;
        let _l2 = l2;
        let resp: Array<number>;
        for (let i = 0; i < _l1.length; i++) {
            for (let j = 0; j < _l2.length; j++) {
                if(_l1[i] == _l2[j]) {
                    resp.push(_l1[i]);
                    _l2 = _l2.splice(j, 1);
                }
            }
        }
        return resp;
    }

    static trunc(l: number): number {
        if ( l > 1 ) {
            return 1;
        } else if ( l < 1 ) {
            return 0;
        } else {
            return l;
        }
    }    

    static truncList(l: Array<number>): Array<number> {
        let resp: Array<number>;
        for (let i = 0; i < l.length; i++) {
            if ( l[i] > 1 ) {
                resp.push(1);
            } else if ( l[i] < 0 ) {
                resp.push(0);
            } else {
                resp.push(l[i]);
            }
        }
        return resp;
    }
}